10/UI/form/section headings html tags

catenglaender opened the pull request:
https://mantis.ilias.de/view.php?id=42226

# Issue

Sections no longer have heading tags.

![image](https://github.com/user-attachments/assets/cfb1a196-8444-4f29-9fd5-19e4f6d2aaed)

# Changes

Sections now assess their nesting level and html tags are rendered accordingly. ~~I also strengthened the visual difference of differently deep nesting levels.~~

UPDATE: @yvseiler and I decided against visual indents for now. The headline size already does a sufficient job to convey nesting and we don't have deeply nested sections yet. We might revisit this once we have the first use cases of triple+ nested sections (if ever)

UPDATED IMAGE:
![image](https://github.com/user-attachments/assets/88941d95-6c3f-49af-bd95-9f8ab1cc62e2)

# ~~Advice needed~~

~~Unit tests can currently only handle the first level expecting a h2. I will discuss a better solution with my colleagues.~~

Unit Tests now account for both the general form tests and a specific triple nested scenario.

catenglaender committed:
10/UI/input/section nested html headings

klees reviewed (commented):
Hi @catenglaender,

LGTM in general, plain, simple and functional.

* [x] Question 1: Did you check with @yvseiler et al regarding the visuals (indentation etc.).
* [x] Question 2: How will this look on small screens?
* [x] Advice: Please check back with @nhaagen to look into unit tests.

Kind regards!

catenglaender commented:
I contacted @yvseiler and we will have a look at it

@nhaagen gave me guidance on how to do the unit tests. The general getFormWrappedHtml() will only be able to test forms with unnested sections for now. I added another Unit Test to test nested headline creation specifically.

UPDATED mobile screenshot with no indents
![image](https://github.com/user-attachments/assets/4e009ec5-ca5e-494a-afc3-0685bae0a17f)

catenglaender commented:
@yvseiler and I decided against visual indents for now. The headline size already does a sufficient job to convey nesting and we don't have deeply nested sections yet. We might revisit this once we have the first use cases of triple+ nested sections (if ever)

Failing unit tests are not related to this PR.

With that, I think, the work here is done for now. Ready for review/merge :slightly_smiling_face:

klees commented:
Hi @thibsy and @Amstutz,

would you please look into this? Tthis is not the most elegant solution (since immutability is broken), but also the violation of the principle seems to be local. Some like the [catamorphism](https://github.com/ILIAS-eLearning/ILIAS/pull/8812) can produce a nicer solution, but then again we are looking to ship a release. So from my POV this solution is ok for the time being. 

Kind regards!

Amstutz commented:
Hi @catenglaender 

Thx a lot for this PR. We believe this to be ready to be merged, except for one minor detail:

- [x] Snake case: Please use snake_case for all variables e.g. [here](https://github.com/ILIAS-eLearning/ILIAS/pull/8889/files#diff-e193eb78940e9c0dc5787b86593505e07ffac4f4242b9d1bee8b1fff730a0cfbR65).

And one thing you can change or explain why not:
- [x] Counting: You start $nesting_level with 2. That seems a bit odd. We understand, why this is done like so. But if done like that, better rename to heading_level. If kept as nesting level, start at 0 and translate in rendering (+2).

Thx a lot

catenglaender commented:
I made the changes :slightly_smiling_face: 
* variables use snake case (hope I got them all)
* nesting level starts with 0
* translation from nesting level to headline level happens in the renderer, so does the cut off at 6 levels deep.

Amstutz commented:
thx a lot!

klees commented:
@Amstutz Is this picked to trunk?

thibsy commented:
@klees it is, depending on where you click in the "referenced commit message" the proper commit is shown :)

klees commented:
![image](https://github.com/user-attachments/assets/05e09fd9-e4c3-4819-a5da-ff32251f9d3b)

This one! Was tricked by the commit message starting with `10/`.
